aws-observability/aws-application-signals-test-framework
# Add A Retry For Unstarted Jobs

*Issue description:*
Github jobs sometime fails to start like this job [here](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10404651182/job/28813573511). This is due to internal error with Github that we have no control over, so we are implementing a retry mechanism for such cases

*Description of changes:*
Currently we have a `-canary.yml` workflow which initializes enablement test run for each region using a matrix. We are adding a second layer between the `-canary.yml` and `e2e-test.yml` workflow called `-retry.yml` that checks if the `e2e-test.yml` workflow failed due to the above mentioned error, and if so, run a new test. 

- Removed references to `e2e`, replaced with `enablement` if needed
- Moved concurrency group to `-retry.yml`, updated concurrency group to have branch name and also manually inputted workflow name. This is because if the different enablement tests are defined in the same workflow, such as when we run tests or in release testing, then the concurrency group start blocking each other and make the overall run slower to finish
- Moved public metric job to `-retry.yml`

*Rollback procedure:*
Revert

*Test run:*
Successful EKS: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10517002486)
Successful EC2: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10517004283)
Successful K8s: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10517001679)

First enablement attempt failed: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10516956484)
Second enablement attempt failed too: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10516286398)
First retry attempt failed: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10516289584)
First enablement attempt and first retry attempt failed: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10516480860)
Both enablement attempts and first retry attempt failed: [Run](https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/10516487372)

- Checked that second jobs run when first jobs fail
- Checked that this logic `validation-result: ${{ needs.java-ec2-asg-attempt-1.outputs.validation-result || needs.java-ec2-asg-attempt-2.outputs.validation-result }}` outputs the success value when either the enablement attempt 1 or attempt 2 succeeds regardless of whether it was publish attempt 1 or 2

<Can we safely revert this commit if needed? If not, detail what must be done to safely revert and why it is needed.>

*Ensure you've run the following tests on your changes and include the link below:*

To do so, create a `test.yml` file with `name: Test` and workflow description to test your changes, then remove the file for your PR. Link your test run in your PR description. This process is a short term solution while we work on creating a staging environment for testing.

NOTE: TESTS RUNNING ON A SINGLE EKS CLUSTER CANNOT BE RUN IN PARALLEL. See the [needs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) keyword to run tests in succession.
- Run Java EKS on `e2e-playground` in us-east-1 and eu-central-2
- Run Python EKS on `e2e-playground` in us-east-1 and eu-central-2
- Run metric limiter on EKS cluster `e2e-playground` in us-east-1 and eu-central-2
- Run EC2 tests in all regions
- Run K8s on a separate K8s cluster (check IAD test account for master node endpoints; these will change as we create and destroy clusters for OS patching)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
